Pcap++/src/KniDevice.cpp

```diff
@@ -0,0 +1,1014 @@
+	m_Device(NULL), m_MBufMempool(NULL)
+{
+	struct rte_kni_ops kni_ops;
+	struct rte_kni_conf kni_conf;
+	m_DeviceInfo.init(conf);
+	m_Requests.thread = NULL;
+	std::memset(&m_Capturing, 0, sizeof(m_Capturing));
+	std::memset(&m_Requests, 0, sizeof(m_Requests));
+
+	if ((m_MBufMempool = create_mempool(mempoolSize, unique, conf.name)) == NULL)
+		return;
+
+	std::memset(&kni_ops, 0, sizeof(kni_ops));
+	std::memset(&kni_conf, 0, sizeof(kni_conf));
+	snprintf(kni_conf.name, RTE_KNI_NAMESIZE, "%s", conf.name);
+	kni_conf.core_id = conf.kthreadCoreId;
+	kni_conf.mbuf_size = MBufRawPacket::MBUF_DATA_SIZE;
+	kni_conf.force_bind = conf.bindKthread ? 1 : 0;
+#if RTE_VERSION >= RTE_VERSION_NUM(18, 2, 0, 0)
+	if (conf.mac != NULL)
+		conf.mac->copyTo((uint8_t*)kni_conf.mac_addr);
+	kni_conf.mtu = conf.mtu;
+#endif
+
+	kni_ops.port_id = conf.portId;
+#if RTE_VERSION >= RTE_VERSION_NUM(17, 11, 0, 0)
+	if (conf.callbacks != NULL)
+	{
+		kni_ops.change_mtu = conf.callbacks->change_mtu;
+		kni_ops.config_network_if = conf.callbacks->config_network_if;
+	#if RTE_VERSION >= RTE_VERSION_NUM(18, 2, 0, 0)
+		kni_ops.config_mac_address = conf.callbacks->config_mac_address;
+		kni_ops.config_promiscusity = conf.callbacks->config_promiscusity;
+	#endif
+	}
+#else
+	if (conf.oldCallbacks != NULL)
+	{
+		kni_ops.change_mtu = conf.oldCallbacks->change_mtu;
+		kni_ops.config_network_if = conf.oldCallbacks->config_network_if;
+	}
+#endif
+
+	m_Device = rte_kni_alloc(m_MBufMempool, &kni_conf, &kni_ops);
+	if (m_Device == NULL)
+	{
+		LOG_ERROR("DPDK have failed to create KNI device %s", conf.name);
+	}
+}
+
+KniDevice::~KniDevice()
+{
+	m_Requests.cleanup();
+	m_Capturing.cleanup();
+	m_DeviceInfo.cleanup();
+	set_kni_device_link_state(m_Device, m_DeviceInfo.name, KniDevice::LINK_DOWN);
+	destroy_kni_device(m_Device, m_DeviceInfo.name);
+}
+
+KniDevice::KniLinkState KniDevice::updateLinkState(KniLinkState state)
+{
+	KniLinkState oldState = set_kni_device_link_state(m_Device, m_DeviceInfo.name, state);
+	if (oldState != KniDevice::LINK_NOT_SUPPORTED && oldState != KniDevice::LINK_ERROR)
+		m_DeviceInfo.link = state;
+	return oldState;
+}
+
+bool KniDevice::handleRequests()
+{
+	return rte_kni_handle_request(m_Device) == 0;
+}
+
+void KniDevice::KniRequests::cleanup()
+{
+	if (thread)
+		thread->cancel();
+	delete thread;
+	thread = NULL;
+	sleepS = sleepNs = 0;
+}
+
+void* KniDevice::KniRequests::runRequests(void* p)
+{
+	KniDevice* device = (KniDevice*)p;
+	struct timespec sleepTime;
+	sleepTime.tv_sec = device->m_Requests.sleepS;
+	sleepTime.tv_nsec = device->m_Requests.sleepNs;
+	struct rte_kni* kni_dev = device->m_Device;
+	for(;;)
+	{
+		nanosleep(&sleepTime, NULL);
+		rte_kni_handle_request(kni_dev);
+	}
+	return NULL;
+}
+
+bool KniDevice::startRequestHandlerThread(long sleepSeconds, long sleepNanoSeconds)
+{
+	if (m_Requests.thread != NULL)
+	{
+		LOG_DEBUG("KNI request thread is already started for device \"%s\"", m_DeviceInfo.name);
```

Comment:
I don't see this as an error. If you want it to be an error i will correct the macro

Reply:
according to the method's documentation: 
> @return true if new thread is started successfully false otherwise

If this methods returns false, the user might want to understand what happened without changing log level to debug

Reply:
Ok. I am now better understand your approach.

Reply:
\+